src/connectivity/all_pairs_all_simple_paths.rs

```diff
@@ -0,0 +1,94 @@
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+use rayon::prelude::*;
+
+use petgraph::algo;
+use petgraph::graph::NodeIndex;
+use petgraph::EdgeType;
+
+use crate::iterators::{AllPairsMultiplePathMapping, MultiplePathMapping};
+use crate::StablePyGraph;
+use retworkx_core::dictmap::*;
+
+pub fn all_pairs_all_simple_paths<Ty: EdgeType + Sync>(
+    graph: &StablePyGraph<Ty>,
+    min_depth: Option<usize>,
+    cutoff: Option<usize>,
+) -> AllPairsMultiplePathMapping {
+    if graph.node_count() == 0 {
+        return AllPairsMultiplePathMapping {
+            paths: DictMap::new(),
+        };
+    } else if graph.edge_count() == 0 {
+        return AllPairsMultiplePathMapping {
+            paths: graph
+                .node_indices()
+                .map(|i| {
+                    (
+                        i.index(),
+                        MultiplePathMapping {
+                            paths: DictMap::new(),
+                        },
+                    )
+                })
+                .collect(),
+        };
+    }
+    let intermediate_min: usize = match min_depth {
+        Some(depth) => depth - 2,
+        None => 0,
+    };
+    let intermediate_cutoff = cutoff.map(|depth| depth - 2);
```

Comment:
raise a `ValueError` if a user provides `min_depth` or `cutoff` less than 2.

Reply:
Done in: https://github.com/Qiskit/retworkx/pull/577/commits/f4490cd901b21978d334eaf15ca69ca39d5e6683